File: UI/travle_core/lib/src/models/tour_response.dart
import 'package:json_annotation/json_annotation.dart';

import 'tour_destination_ref.dart';
import 'tour_schedule_response.dart';

part 'tour_response.g.dart';

/// A bookable tour in the single shape every read path returns — list card,
/// "tours visiting a destination", and detail (mirrors the backend `TourResponse`).
/// Only [primaryThumbnail] carries bytes on the list path (the ordered-first
/// destination's thumbnail, reused as the cover). [destinations] and [schedules]
/// are populated on the detail read only, so list payloads stay light.
@JsonSerializable()
class TourResponse {
  TourResponse({
    required this.id,
    required this.name,
    required this.description,
    required this.durationMinutes,
    required this.pricePerPerson,
    required this.capacity,
    this.entranceFeesPerPerson = 0,
    required this.tourTypeId,
    required this.organizerId,
    required this.isActive,
    required this.destinationCount,
    required this.upcomingScheduleCount,
    required this.createdAt,
    this.hasUnavailableDestination = false,
    this.averageRating = 0,
    this.reviewCount = 0,
    this.isFavorite = false,
    this.tourTypeName,
    this.organizerName,
    this.nextDepartureAt,
    this.timeZoneId = 'UTC',
    this.primaryThumbnail,
    this.primaryThumbnailContentType,
    this.destinations = const [],
    this.schedules,
    this.modifiedAt,
  });

  final int id;

  final String name;
  final String description;

  final int durationMinutes;
  final double pricePerPerson;

  /// Sum of the visited destinations' entrance fees, per person (KM), paid on-site
  /// and never part of the Travle charge. 0 when no stop has a fee.
  final double entranceFeesPerPerson;

  /// Default group size seeded into new schedules; each slot may override it.
  final int capacity;

  final int tourTypeId;
  final String? tourTypeName;

  final int organizerId;
  final String? organizerName;

  final bool isActive;

  /// True when a stop left the approved catalogue (edited back to moderation / rejected). Such a tour is
  /// hidden from travelers; on the organizer's own list it drives a "Temporarily unavailable" badge.
  final bool hasUnavailableDestination;

  /// Average of the tour's non-removed reviews (0 when it has none). Computed on read.
  final double averageRating;

  /// Number of non-removed reviews behind [averageRating].
  final int reviewCount;

  /// Whether the current user has this tour in their favorites (heart state).
  final bool isFavorite;

  final int destinationCount;
  final int upcomingScheduleCount;
  final DateTime? nextDepartureAt;

  /// IANA time-zone id (e.g. "Europe/Sarajevo") the tour's event times display in (UTC instants on the wire).
  @JsonKey(defaultValue: 'UTC')
  final String timeZoneId;

  /// Base64 cover thumbnail for list cards (null when no stop has an image).
  final String? primaryThumbnail;
  final String? primaryThumbnailContentType;

  /// Ordered stops — populated on the detail read only (empty in lists).
  final List<TourDestinationRef> destinations;

  /// Upcoming Active schedules with live free-seat counts — detail read only.
  final List<TourScheduleResponse>? schedules;

  final DateTime createdAt;
  final DateTime? modifiedAt;

  factory TourResponse.fromJson(Map<String, dynamic> json) =>
      _$TourResponseFromJson(json);

  Map<String, dynamic> toJson() => _$TourResponseToJson(this);
}

File: Backend/Travle.Model/Responses/TourResponse.cs
namespace Travle.Model.Responses
{
    /// <summary>
    /// A bookable tour, in the single shape used by every read path — the paginated list, the "tours
    /// visiting a destination" section, and the detail view. Reference fields are flattened to names
    /// (never raw ids on screen). Only <see cref="PrimaryThumbnail"/> carries bytes on the list path
    /// (the ordered-first destination's thumbnail, reused as the tour cover — tours have no images of
    /// their own); <see cref="Destinations"/> and <see cref="Schedules"/> are populated on the detail
    /// path only, so list payloads stay light (§8.2 / rule 12).
    /// </summary>
    public class TourResponse
    {
        public int Id { get; set; }

        public string Name { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;

        public int DurationMinutes { get; set; }
        public decimal PricePerPerson { get; set; }

        /// <summary>
        /// Sum of the visited destinations' informative entrance fees (KM), <b>per person</b> — money the
        /// traveler pays on-site, never part of the Travle charge. 0 when no stop has a fee. Shown as a
        /// "bring around X" guide, not an authoritative total.
        /// </summary>
        public decimal EntranceFeesPerPerson { get; set; }

        /// <summary>Default group size seeded into new schedules; each slot may override it.</summary>
        public int Capacity { get; set; }

        public int TourTypeId { get; set; }
        public string? TourTypeName { get; set; }

        public int OrganizerId { get; set; }
        public string? OrganizerName { get; set; }

        public bool IsActive { get; set; }

        /// <summary>Number of destinations this tour visits (shown on cards without loading the stops).</summary>
        public int DestinationCount { get; set; }

        /// <summary>Count of Active, future schedules (a "sold-out"/"no upcoming dates" hint on cards).</summary>
        public int UpcomingScheduleCount { get; set; }

        /// <summary>Start of the next Active, future schedule, if any (for a "next departure" chip).</summary>
        public DateTime? NextDepartureAt { get; set; }

        /// <summary>The ordered-first destination's thumbnail, used as the tour cover on list cards.</summary>
        public byte[]? PrimaryThumbnail { get; set; }
        public string? PrimaryThumbnailContentType { get; set; }

        /// <summary>Ordered stops the tour visits. Populated on the detail read only (empty in lists).</summary>
        public List<TourDestinationRef> Destinations { get; set; } = new List<TourDestinationRef>();

        /// <summary>Upcoming Active schedules with live free-seat counts. Detail read only (null in lists).</summary>
        public List<TourScheduleResponse>? Schedules { get; set; }

        public DateTime CreatedAt { get; set; }
        public DateTime? ModifiedAt { get; set; }
    }
}
